joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h

```diff
@@ -811,6 +688,151 @@ setHoldPosition(const ros::Time& time, RealtimeGoalHandlePtr gh)
+  // next control cycle, leaving the current cycle without a valid trajectory.
+
+  // Update current state and state error
+  for (unsigned int i = 0; i < joints_.size(); ++i)
+  {
+    current_state_.position[i] = joints_[i].getPosition();
+    current_state_.velocity[i] = joints_[i].getVelocity();
+    // There's no acceleration data available in a joint handle
+
+    typename TrajectoryPerJoint::const_iterator segment_it = sample(curr_traj[i], time_data.uptime.toSec(), desired_joint_state_);
+    if (curr_traj[i].end() == segment_it)
+    {
+      // Non-realtime safe, but should never happen under normal operation
+      ROS_ERROR_NAMED(name_,
+                      "Unexpected error: No trajectory defined at current time. Please contact the package maintainer.");
+      return;
+    }
+    desired_state_.position[i] = desired_joint_state_.position[0];
+    desired_state_.velocity[i] = desired_joint_state_.velocity[0];
+    desired_state_.acceleration[i] = desired_joint_state_.acceleration[0]; ;
+
+    state_joint_error_.position[0] = angles::shortest_angular_distance(current_state_.position[i],desired_joint_state_.position[0]);
+    state_joint_error_.velocity[0] = desired_joint_state_.velocity[0] - current_state_.velocity[i];
+    state_joint_error_.acceleration[0] = 0.0;
+
+    state_error_.position[i] = angles::shortest_angular_distance(current_state_.position[i],desired_joint_state_.position[0]);
+    state_error_.velocity[i] = desired_joint_state_.velocity[0] - current_state_.velocity[i];
+    state_error_.acceleration[i] = 0.0;
+
+    //Check tolerances
+    const RealtimeGoalHandlePtr rt_segment_goal = segment_it->getGoalHandle();
+    if (rt_segment_goal && rt_segment_goal == rt_active_goal_)
+    {
+      // Check tolerances
+      if (time_data.uptime.toSec() < segment_it->endTime())
+      {
+        // Currently executing a segment: check path tolerances
+        const SegmentTolerancesPerJoint<Scalar>& joint_tolerances = segment_it->getTolerances();
+        if (!checkStateTolerancePerJoint(state_joint_error_, joint_tolerances.state_tolerance))
+        {
+          if (verbose_)
+          {
+            ROS_ERROR_STREAM_NAMED(name_,"Path tolerances failed for joint: " << joint_names_[i]);
+            checkStateTolerancePerJoint(state_joint_error_, joint_tolerances.state_tolerance, true);
+          }
+          rt_segment_goal->preallocated_result_->error_code =
+          control_msgs::FollowJointTrajectoryResult::PATH_TOLERANCE_VIOLATED;
+          rt_segment_goal->setAborted(rt_segment_goal->preallocated_result_);
+          rt_active_goal_.reset();
+          successful_joint_traj_.reset();
+        }
+      }
+      else if (segment_it == --curr_traj[i].end())
+      {
+        if (verbose_)
+          ROS_DEBUG_STREAM_THROTTLE_NAMED(1,name_,"Finished executing last segment, checking goal tolerances");
+
+        // Controller uptime
+        const ros::Time uptime = time_data.uptime;
```

Comment:
Why suddenly a copy? Why not a const reference before joint-loop?

Furthermore: At every call to `uptime` in this function we see: `...uptime.toSec()`. Why not simply do:
```
const double uptime_sec {time_data.uptime.toSec()};
```
before the joint-loop?

Reply:
!!! Changed behavior:
The original lines looked like this:
```
const ros::Time uptime = time_data_.readFromRT()->uptime;
```
But now you use the local variable `time_data` which is:
```
time_data.uptime = time_data_.readFromRT()->uptime + period
```
Is this change intended?